sapicore/data/sampling/__init__.py

```diff
@@ -0,0 +1,136 @@
+
+        n: int or float
+            Specifies the number or fraction of samples to draw from each group.
+            While the former is more convenient for equal (disproportionate) sampling and the latter
+            for stratified (proportionate) sampling, integer or float values can be used with either one.
+
+        """
+        # keep track of global indices for subset selection.
+        frame["index"] = frame.index
+        grouped = frame.groupby(group_keys, group_keys=False)
+
+        if self.stratified:
+            # convert `n` to fraction if need be.
+            if isinstance(n, int):
+                frac = len(frame["index"].tolist()) * n
+
+            # perform stratified sampling of `frac` out of every group.
+            subset = grouped.apply(lambda x: x.sample(frac=frac, replace=self.replace))
+
+        else:
+            # convert `n` to integer if need be.
+            if isinstance(n, float):
+                n = int(n * len(frame["index"].tolist()))
+
+            subset = grouped.apply(lambda x: x.sample(n, replace=self.replace))
+
+        return subset["index"].tolist()
+
+
+class CV:
+    """Cross validation base class.
+
+    Parameter
+    ---------
+    data: Data
+        The :class:`data.Data` object for which to generate the cross validator.
+
+    cross_validator: BaseCrossValidator
+        A scikit-learn :class:`BaseCrossValidator` object.
+
+    label_keys: str or list of str, optional
+        One or more descriptor label names. Required for some CV schemes (e.g., for StratifiedKFold, it would be
+        the descriptor whose class label counts should be consistent with their proportions in the full sample).
+
+    group_key: str, optional
+        A grouping descriptor key. Required for some CV schemes (e.g., GroupKFold, for generating batch-like
+        folds with non-overlapping groups w.r.t. some descriptor).
+
+    Note
+    ----
+    Some external libraries extend the cross validator offerings of scikit-learn.
+    This implementation is compatible with any class derived from BaseCrossValidator.
+
+    """
+
+    def __init__(
+        self,
+        data,
+        cross_validator: BaseCrossValidator,
+        label_keys: str | list[str] = None,
+        group_key: str = None,
+        **kwargs
+    ):
+
+        self.data = data
+        self.cross_validator = cross_validator
+
+        if isinstance(label_keys, str):
+            self.labels = self.data.descriptors[label_keys].labels
+        else:
+            # if multiple stratification labels were provided,
+            self.labels = list(zip(*[self.data.descriptors[i].labels for i in label_keys]))
+
+        self.groups = self.data.descriptors[group_key].labels if group_key else None
+
+    def __iter__(self):
+        self.index = 0
+        self.splitter = self.cross_validator.split(X=self.labels, y=self.labels, groups=self.groups)
+
+        return self.splitter
```

Comment:
You really should be creating copies of this class for every call to this function. Because it's normal to create multiple iterators (e.g. with `it1 = iter(cv); it2 = iter(cv)`) and iterate them simultaneously or in random order or whatever. So the assumption that you'll only have on iterator at once called in serial order is not valid. Generally `__iter__` needs to return a new iterator for every call.

Reply:
> Generally __iter__ needs to return a new iterator for every call.

Though it's not obvious, that's exactly what's happening here.

To convince yourself, try:

```python3
    x = Data(buffer=torch.rand(4, 4), metadata=Metadata(AxisDescriptor("a", [1, 2, 3, 4])))
    cv = CV(x, cross_validator=KFold(n_splits=2), label_keys="a")

    temp1 = iter(cv)
    temp2 = iter(cv)

    for _ in range(2):
        print(next(temp1))
        print(next(temp2))
```
This is happening because cross_validator.split returns a generator. Just Sklearn stuff.

Reply:
I was thinking, a simpler way of doing stuff, if index is the only thing that changes in iterations is the following pattern:


```py

class WaveIterable:

    def __init__(self, wave):
        self.index = 0
        self.wave = wave

    def __next__(self):
        value = self.wave._next(self.index)
        self.index += 1
        return value


class Wave:

    def __iter__(self):
        return WaveIterable(self)

    def _next(self, index):
        # compute wave at current time point from given components.
        value = torch.zeros(1, dtype=torch.float, device=DEVICE)

        if self.valid:
            for i in range(len(self.frequencies)):
                amp_series = self.amplitudes[i]
                time = index * (1.0 / self.sampling_rate)

                if self.phase_amplitude_coupling is not None:
                    # add sine variation to base amplitude.
                    temp = self.amplitudes[i] * torch.sin(
                        2 * torch.pi * self.phase_amplitude_coupling[i] * time + self.phase_shifts[i]
                    )
                    amp_series = amp_series + temp

                # add PAC oscillation to combined wave.
                value = value + amp_series * torch.sin(2 * torch.pi * self.frequencies[i] * time + self.phase_shifts[i])

        return value
```

This keeps the wave specific stuff in the wave class with the iterator only calling into that class.

Reply:
Right,  `__next__` here is an unnecessary leftover from a previous version and does lead to the problem you stated (wraps sklearn's next but only for the current self.splitter). 

I'll implement your `WaveIterable` fix too, either in this round or the next.